src/data/qpf/multivariate/constructions/cofix.lean

```diff
@@ -0,0 +1,433 @@
+  dsimp [liftr'], split; refl,
+end
+
+open function
+
+theorem liftr_map_last [is_lawful_mvfunctor F] {α : typevec n} {ι ι'}
+  (R : ι' → ι' → Prop) (x : F (α ::: ι)) (f g : ι → ι')
+  (hh : ∀ x : ι, R (f x) (g x)) :
+  liftr' (rel_last' _ R) ((id ::: f) <$$> x) ((id ::: g) <$$> x) :=
+let h : ι → { x : ι' × ι' // uncurry R x } := λ x, ⟨ (f x,g x), hh x ⟩ in
+let b : α ::: ι ⟹ _ := @diag_sub n α ::: h,
+    c : subtype_ α.repeat_eq ::: {x // uncurry R x} ⟹
+        (λ (i : fin2 (n)), {x // of_repeat (α.rel_last' R i.fs x)}) ::: subtype (uncurry R) :=
+      of_subtype _ ::: id
+in
+have hh : subtype_val _ ⊚ to_subtype _ ⊚ from_append1_drop_last ⊚ c ⊚ b =
+          (id ::: f ⊗' id ::: g) ⊚ prod.diag,
+  by { dsimp [c,b],
+       apply eq_of_drop_last_eq,
+       { dsimp,
+         simp only [prod_map_id, drop_fun_prod, drop_fun_append_fun, drop_fun_diag, id_comp,
+                    drop_fun_to_subtype],
+         erw [to_subtype_of_subtype_assoc,id_comp],
+         clear_except,
+         ext i x : 2, induction i,
+         refl,  apply i_ih, },
+       simp only [h, last_fun_from_append1_drop_last, last_fun_to_subtype, last_fun_append_fun,
+                  last_fun_subtype_val, comp.left_id, last_fun_comp, last_fun_prod],
+       dsimp, ext1, refl },
+liftr_map _ _ _ _ (to_subtype _ ⊚ from_append1_drop_last ⊚ c ⊚ b) hh
+
+theorem liftr_map_last' [is_lawful_mvfunctor F] {α : typevec n} {ι}
+  (R : ι → ι → Prop) (x : F (α ::: ι)) (f : ι → ι)
+  (hh : ∀ x : ι, R (f x) x) :
+  liftr' (rel_last' _ R) ((id ::: f) <$$> x) x :=
+begin
+  have := liftr_map_last R x f id hh,
+  rwa [append_fun_id_id,mvfunctor.id_map] at this,
+end
+
+end
+
+lemma cofix.abs_repr {α} (x : cofix F α) :
+  quot.mk _ (cofix.repr x) = x :=
+begin
+  let R := λ x y : cofix F α,
+    cofix.abs (cofix.repr y) = x,
+  refine cofix.bisim₂ R _ _ _ rfl,
+  clear x, rintros x y h, dsimp [R] at h, subst h,
+  dsimp [cofix.dest,cofix.abs],
+  induction y using quot.ind,
+  simp only [cofix.repr, M.dest_corec, abs_map, abs_repr],
+  conv { congr, skip, rw cofix.dest },
+  dsimp, rw [mvfunctor.map_map,mvfunctor.map_map,← append_fun_comp_id,← append_fun_comp_id],
+  let f : α ::: (P F).M α ⟹ subtype_ (α.rel_last' R) :=
+    split_fun diag_sub (λ x, ⟨(cofix.abs (cofix.abs x).repr, cofix.abs x),_⟩),
+  refine liftr_map _ _ _ _ f _,
+  { simp only [←append_prod_append_fun, prod_map_id],
+    apply eq_of_drop_last_eq,
+    { dsimp, simp only [drop_fun_diag],
+      erw subtype_val_diag_sub,
+    },
+    ext1,
+    simp only [cofix.abs, prod.mk.inj_iff, prod_map, function.comp_app, last_fun_append_fun,
+               last_fun_subtype_val, last_fun_comp, last_fun_split_fun],
+    dsimp [drop_fun_rel_last,last_fun,prod.diag],
+    split; refl, },
+  dsimp [rel_last',split_fun,function.uncurry,R],
+  refl,
+end
+
+section tactic
+
+setup_tactic_parser
+open tactic
+omit q
+
+/-- tactic for proof by bisimulation -/
```

Comment:
Is this supposed to be used by anyone? If so, the tactic documentation should give the syntax and an example. Also, why the subscripted 2?

Reply:
This could be useful. I think the implementation will need to provide better tactics though. 

As for the subscript, to be usable, it has to be in the `tactic.interactive` namespace but the univariate `qpf` has a similar tactic so I needed to make the name globally unique. Maybe I should not put them in `tactic.interactive` and just use them locally.